From 64b55b742a7580109494e3c9431c1d61800352ca Mon Sep 17 00:00:00 2001
From: Raphael Mounier <mounierr07@gmail.com>
Date: Wed, 17 May 2023 19:53:52 +0200
Subject: [PATCH] core : fix blanking off in huawei device

For some unknown reason, the call to the gr_fb_blank(blank) function does not work on Huawei with kirin processors. The cause is at the level of the minui library in the call to the function which turns off the screen but cannot turn it back on

ret = ioctl(fb_fd, FBIOBLANK, blank? FB_BLANK_POWERDOWN: FB_BLANK_UNBLANK)

 which does not work.

The call to this function is therefore replaced by a simple filling of the empty screen.

Change-Id: Id7d4b476495cb063ded3779480741015afcff3ea
---
 healthd/healthd_draw.cpp | 43 +++++++++++++++++++++++++++++++++++++++-
 1 file changed, 42 insertions(+), 1 deletion(-)

diff --git a/healthd/healthd_draw.cpp b/healthd/healthd_draw.cpp
index 7c7931944..aa6558b10 100644
--- a/healthd/healthd_draw.cpp
+++ b/healthd/healthd_draw.cpp
@@ -17,6 +17,7 @@
 #include <android-base/stringprintf.h>
 #include <batteryservice/BatteryService.h>
 #include <cutils/klog.h>
+#include <cutils/properties.h>
 
 #include "healthd_draw.h"
 
@@ -24,6 +25,8 @@
 #include "charger.sysprop.h"
 #endif
 
+#define HARDWARE_MODEL "ro.hardware"
+
 #define LOGE(x...) KLOG_ERROR("charger", x);
 #define LOGW(x...) KLOG_WARNING("charger", x);
 #define LOGV(x...) KLOG_DEBUG("charger", x);
@@ -73,6 +76,7 @@ HealthdDraw::HealthdDraw(animation* anim)
         (res = gr_init_font(anim->text_clock.font_file.c_str(), &anim->text_clock.font)) < 0) {
         LOGE("Could not load time font (%d)\n", res);
     }
+
     if (!anim->text_percent.font_file.empty() &&
         (res = gr_init_font(anim->text_percent.font_file.c_str(), &anim->text_percent.font)) < 0) {
         LOGE("Could not load percent font (%d)\n", res);
@@ -95,8 +99,45 @@ void HealthdDraw::redraw_screen(const animation* batt_anim, GRSurface* surf_unkn
 }
 
 void HealthdDraw::blank_screen(bool blank, int drm) {
+
     if (!graphics_available) return;
-    gr_fb_blank(blank, drm);
+
+    bool bmulti=gr_has_multiple_connectors();
+
+    if (bmulti && (drm==1)) {
+        KLOG_WARNING("charger", "minui graphic backend don't support multi-connector for blank screen\n");
+    }
+
+	// blank don't work on huawei device, use clear_screen()
+    bool is_kirin=false;
+    char prop_hardware[PROPERTY_VALUE_MAX] = {};
+
+    if (property_get(HARDWARE_MODEL, prop_hardware, "") > 0) {
+		if (!strcmp(prop_hardware,"hi3660")
+			|| !strcmp(prop_hardware,"hi3670")
+			|| !strcmp(prop_hardware,"hi6250")
+			|| !strcmp(prop_hardware,"kirin"))
+		{
+			LOGV("Kirin Huawei found");
+			is_kirin=true;
+		}
+	}
+
+	if (is_kirin) {
+		if (blank==true) {
+			LOGV("Kirin - Just clear screen");
+			clear_screen();
+			gr_flip();
+		}
+		else {
+			LOGV("Kirin - Do not touch screen");
+		}
+	}
+	else {
+		LOGV("Blank screen with minui api");
+		gr_fb_blank(blank, drm);
+	}
+
 }
 
 // support screen rotation for foldable phone
-- 
2.25.1

